flink-tests/src/test/java/org/apache/flink/test/checkpointing/ChangelogPeriodicMaterializationITCase.java

```diff
@@ -0,0 +1,734 @@
+
+        public PreHandlingSource() {
+            this.completedCheckpointNum = new AtomicInteger();
+        }
+
+        public static void initSourceData(int totalNum) {
+            sourceList = new ArrayList<>(totalNum);
+            for (int i = 0; i < totalNum; i++) {
+                sourceList.add(ThreadLocalRandom.current().nextInt(totalNum));
+            }
+        }
+
+        public static Map<Integer, Integer> getActualResult() {
+            return sourceList.stream()
+                    .collect(
+                            Collectors.toConcurrentMap(
+                                    Function.identity(), element -> 1, Integer::sum));
+        }
+
+        @Override
+        public void snapshotState(FunctionSnapshotContext context) throws Exception {
+            lastSentStored.update(Arrays.asList(currentIndex, completedCheckpointNum.get()));
+        }
+
+        @Override
+        public void initializeState(FunctionInitializationContext context) throws Exception {
+            lastSentStored =
+                    context.getOperatorStateStore()
+                            .getListState(new ListStateDescriptor<>("counter", Integer.class));
+            if (context.isRestored()) {
+                Iterable<Integer> indexAndCheckpointNum = lastSentStored.get();
+                currentIndex = get(indexAndCheckpointNum, 0);
+                completedCheckpointNum.compareAndSet(0, get(indexAndCheckpointNum, 1));
+            }
+        }
+
+        @Override
+        public void run(SourceContext<Integer> ctx) throws Exception {
+            while (!isCanceled && currentIndex < sourceList.size()) {
+                preHandle();
+                synchronized (ctx.getCheckpointLock()) {
+                    if (!isCanceled && currentIndex < sourceList.size()) {
+                        int currentElement = sourceList.get(currentIndex++);
+                        ctx.collect(currentElement);
+                    }
+                }
+            }
+        }
+
+        protected void preHandle() throws Exception {
+            // do nothing by default
+        }
+
+        protected void waitUntilFalse(SerializableBooleanSupplier supplier)
+                throws InterruptedException {
+            while (supplier.getAsBoolean()) {
+                Thread.sleep(10);
+            }
+        }
+
+        protected void throwArtificialFailure() throws Exception {
+            throw new Exception("Artificial Failure");
+        }
+
+        @Override
+        public void cancel() {
+            isCanceled = true;
+        }
+
+        @Override
+        public void notifyCheckpointComplete(long checkpointId) throws Exception {
+            completedCheckpointNum.getAndIncrement();
+        }
+    }
+
+    /** A sink puts the result of WordCount into a Map using for validating. */
+    private static class CollectionSink implements SinkFunction<Tuple2<Integer, Integer>> {
```

Comment:
Can't we use `DataStreaam.executeAndCollect` or accumulators instead of a new `CollectionSink`? 
That would allow to avoid static thread-safe fields and adding a new class.

Reply:
The new CollectionSink is just used to check the correctness of the result.